### ISSUE TITLE
Kotlin native library soname

### ISSUE THREAD

**Issue by arpie42** [JetBrains/kotlin-native#2526]
I'm trying to create a dynamic library in kotlin-native for use within an android project.  Everything appears to compile and link correctly (using gradle) and an apk is pushed to my android device.  But when I run the app, I get an error on my android device :
System.err: dlopen failed: library "/files/Projects/BiaB/one/app/build/intermediates/ndkBuild/debug/obj/local/armeabi-v7a/libgamelib.so"
Of course that path is wrong - it is the local build path on my development machine.

As far as I can see, this is the same problem (but for Go rather than Kotlin) :
https://github.com/golang/go/issues/17807

I am compiling and linking the k/n library independently of the android build, then including it in the android build as a prebuilt shared library.  So as far as I understand, I need to set the 'soname' of the library during linkage of the library itself (before it gets anywhere near my android build).  Is that correct?  In which case, is there a simple flag I can add to -linker-options that will set the soname?

Or is there some other way to sort this out?

**Comment by alexander-gorshenev** [comment]
First of all, `-Xverbose-phases=linker` compiler option allows one to take a look at what is being used for linker invocation.

Another useful option is `-Xtemporary-files-dir=somedir` to preserve temporary files in `somedir` directory. That allows to reproduce the link step manually.

Using these option I see that for `android_arm64` target the compiler just invokes `clang` on the resulting `.o` file. That suggests asking clang to pass the proper linker flag using `-linker-options "-Wl,-soname,XYZ"`.

Now, to check the result is good, let's inspect the `.so` file with the target specific `objdump` utility from the kotlin/native installation:

`$ ~/.konan/dependencies/target-toolchain-21-osx-android_arm64/aarch64-linux-android/bin/objdump -x ./libdynamic.so | grep SONAME`

produces the desired

`  SONAME               XYZ`

**Comment by arpie42** [comment]
That's all really useful stuff.  A great big thank you for explaining your thought process as well as providing a solution.
I did try that linker-options flag but, due to not being sure how to use it, I think I mistook that Wl for a W1 (stupid lower case L!).

The actual build is being managed by the following konanc command, as called by the gradle multiplatform plugin :
konanc -g -ea -target android_arm32 -p dynamic -o /files/Projects/mylib/build/bin/nandroid32/main/debug/shared/libgamelib.so -linker-options -Wl,-soname,libgamelib -Xmulti-platform /files/Projects/mylib/src/commonMain/kotlin/main.kt -Xcommon-sources=/files/Projects/mylib/src/commonMain/kotlin/main.kt

I have manually added those -linker-options in the gradle.build file.

Just a thought but would it be worth hardwiring konanc to automatically pass the -soname flag to the linker, using the filename of the -o flag, whenever -p is set to dynamic? I'm sure somebody wiser than me can explain why it would be a bad idea?
